documentation/general/dotnet-run-file.md

```diff
@@ -0,0 +1,266 @@
+(an error is reported if it points to a directory instead).
+Then the build ignores other entry-point files.
+
+Thanks to this, it is possible to have a structure like
+```
+App/Util.cs
+App/Program1.cs
+App/Program2.cs
+```
+where either `Program1.cs` or `Program2.cs` can be run and both of them have access to `Util.cs`.
+
+In this case, there are multiple implicit projects
+(and during [grow up](#grow-up), multiple project files are materialized
+and the original C# files are moved to the corresponding project subdirectories):
+```
+App/Shared/Util.cs
+App/Shared/Shared.csproj
+App/Program1/Program1.cs
+App/Program1/Program1.csproj
+App/Program2/Program2.cs
+App/Program2/Program2.csproj
+```
+
+The generated folders might need to be named differently to avoid clashes with existing folders.
+
+The entry-point projects (`Program1` and `Program2` in our example)
+have a `ProjectReference` to the non-entry-point `Shared` project.
+
+Unless the [artifacts output layout][artifacts-output] is used (which is recommended),
+those implicit projects mean that build artifacts are placed under those implicit directories
+even though they don't exist on disk prior to build:
+```
+App/Shared/bin/
+App/Shared/obj/
+App/Program1/bin/
+App/Program1/obj/
+App/Program2/bin/
+App/Program2/obj/
+```
+
+## Package references
+
+It is possible to specify NuGet package references via the `#r` directive.
+
+```cs
+#r "nuget: Newtonsoft.Json, 13.0.1"
+```
+
+The C# language needs to be updated to ignore these directives (instead of failing the compilation).
+See [the corresponding language proposal][pound].
+
+If these directives were limited by the language to only appear near the top of the file (similar to `#define` directives),
+the dotnet CLI could be more efficient in searching for them.
+
+It should be also possible to look for these directives from the dotnet CLI via a regex instead of parsing the whole C# file via Roslyn.
+
+We can also consider reporting an error or warning if `#r` directives are spread among multiple files,
+to encourage users to keep them in a single place for clarity.
+Furthermore, if the directives are limited to just one file per virtual project,
+it is more efficient to search for them by the dotnet CLI
+(after the initial search, we can only look at a couple of files on re-runs).
+For example, users could put their `#r`s into `Program1.cs`, `Program2.cs`, and `Util.cs`, but not `Util2.cs`.
+
+During [grow up](#grow-up), `#r` directives are removed from the `.cs` files and turned into `<PackageReference>` elements in the corresponding `.csproj` files.
+For project-based programs, `#r` directives are an error.
+
+## Shebang
+
+Along with `#r`, the language can also ignore `#!` which could be then used for [shebang][shebang] support.
+
+```cs
+#!/usr/bin/dotnet run
+Console.WriteLine("Hello");
+```
+
+It might be beneficial to also ship `dotnet-run` binary
```

Comment:
Another idea is to drop a `cs` executable so that simple `#!/usr/bin/env cs` would work, along of course with being able to just run `cs app.cs` on a prompt.

IIUC `#!/usr/bin/dotnet run` should work in all shells right? Just means that `dotnet` has to be in `/usr/bin`

Reply:
Would that work in a windows environment?

Reply:
> Another idea is to drop a `cs` executable so that simple `#!/usr/bin/env cs` would work, along of course with being able to just run `cs app.cs` on a prompt.

Also `dotnet file.dll` is apparently a thing, so we could perhaps add a support for `dotnet file.cs` as well.

> IIUC `#!/usr/bin/dotnet run` should work in all shells right? Just means that `dotnet` has to be in `/usr/bin`

Yes, that's correct.

> Would that work in a windows environment?

I don't think shebangs can be used on Windows.

Reply:
`dotnet foo.cs` would require support in the native dotnet host, not just the managed SDK CLI. I don't think we should go down that path at this time, the host is much harder to change.